app.py

```diff
@@ -33,36 +41,159 @@ def add():
             if not link_exists:
                 break
 
-        link = Link(long_url=form.long_url.data, short_url=short_url)
+        if (current_user.is_authenticated):
+            link = Link(long_url=form.long_url.data, short_url=short_url, user_id = current_user.id)
+        else:
+            link = Link(long_url=form.long_url.data, short_url=short_url)
+
         db.session.add(link)
         db.session.commit()
-    return redirect(url_for('stats', short_url=short_url))
 
+    if current_user.is_authenticated:
+        return redirect(url_for('index'))
+    
+    return redirect(url_for('info', short_url=short_url))
 
-@app.route("/stats/<short_url>")
-def stats(short_url):
-    form =LinkForm()
+
+@app.route("/<short_url>/info")
+def info(short_url):
     link = Link.query.filter_by(short_url=short_url).first()
-    return render_template("stats.html", form=form, link=link)
+
+    if link == None:
+        abort(404)
+    elif link.owner and link.owner != current_user:
+        abort(401)
+
+    form =LinkForm()
+
+    return render_template("info.html", form=form, link=link)
 
 
 @app.route("/<short_url>")
 def redirect_url(short_url):
     link = Link.query.filter_by(short_url=short_url).first()
 
     if link:
+        link.no_of_clicks += 1
+        db.session.commit()
         return redirect(link.long_url)
     else:
         return redirect(url_for("index"))
 
 
+@app.route('/logout')
+def logout():
+    logout_user()
+    return redirect(url_for('index'))
+
+
+@app.route("/authorize/<provider>")
+def oauth2_authorize(provider):
+    if current_user.is_authenticated:
+        return redirect(url_for('index'))
+
+    provider_data = current_app.config['OAUTH2_PROVIDERS'].get(provider)
+    if provider_data is None:
+        abort(404)
+
+    session['oauth2_state'] = secrets.token_urlsafe(16)
+
+    qs = urlencode({
+        'client_id': provider_data['client_id'],
+        'redirect_uri': url_for('oauth2_callback', provider=provider,
+                                _external=True),
+        'response_type': 'code',
+        'scope': ' '.join(provider_data['scopes']),
+        'state': session['oauth2_state'],
+    })
+
+    return redirect(provider_data['authorize_url'] + '?' + qs)
+
+
+@app.route('/callback/<provider>')
+def oauth2_callback(provider):
+    if current_user.is_authenticated:
+        return redirect(url_for('index'))
+
+    provider_data = current_app.config['OAUTH2_PROVIDERS'].get(provider)
+    if provider_data is None:
+        abort(404)
+
+    if 'error' in request.args:
+        return redirect(url_for('index'))
+
+    if request.args['state'] != session.get('oauth2_state'):
```

Comment:
request.args['state'] will throw a KeyError if None. If request.args['state'] == "None" and session.get('oauth2_state') == None it wont match due to different data types. Added a check for request.args['state'] == None just in case it's possible to get around the KeyError somehow.